Add more command line arguments

anish-mudaraddi opened the pull request:
simple change to add command line arguments to specify where to save output files and log files etc

```
muspinsim --help

usage: __main__.py [-h] [-o OUT_DIR] [-l LOG_PATH] [-f FITREPORT_PATH]
                   input_file

Muspinsim - A program designed to carry out spin dynamics calculations for
muon science experiments

positional arguments:
  input_file            filepath to muspinsim specially formatted text file
                        specifying simulation input parameters.

options:
  -h, --help            show this help message and exit
  -o OUT_DIR, --out-dir OUT_DIR
                        folder to store the output .dat files
  -l LOG_PATH, --log-path LOG_PATH
                        filepath to store simulation logs
  -f FITREPORT_PATH, --fitreport-path FITREPORT_PATH
                        filepath to store fit report if fitting parameters
                        given
```

elichad commented on muspinsim/__main__.py:
```diff
@@ -72,20 +116,30 @@ def main(use_mpi=False):
-            fitter.write_report()
+            if args.fitreport_path:
+                rep_path = check_dir_path(os.path.dirname(args.fitreport_path))
+                rep_fname = os.path.basename(args.fitreport_path)
```
rep_fname needs some default too – running without this arg raises `UnboundLocalError: local variable 'rep_fname' referenced before assignment`

elichad commented on muspinsim/__main__.py:
```diff
@@ -24,29 +24,73 @@
-        logfile = "{0}.log".format(os.path.splitext(args.input_file)[0])
+        if args.log_path:
+            # check if directory exists, if not create it
+            logfile = "{0}/{1}".format(
+                check_dir_path(os.path.dirname(args.log_path)),
+                os.path.basename(args.log_path),
```
We shouldn't assume that a user will always enter a path that includes directories. Using `-l test.log` raises `NotADirectoryError`

elichad commented on muspinsim/__main__.py:
```diff
@@ -72,20 +116,30 @@ def main(use_mpi=False):
         if mpi.is_root:
-            fitter.write_report()
+            if args.fitreport_path:
+                rep_path = check_dir_path(os.path.dirname(args.fitreport_path))
```
Same issue as the log file here - if the user doesn't include a directory this raises `NotADirectoryError`

elichad commented on muspinsim/__main__.py:
```diff
@@ -24,29 +24,73 @@
 LOGFORMAT = "[%(levelname)s] [%(threadName)s] [%(asctime)s] %(message)s"
 
 
+def check_dir_path(string):
```
This function name could be clearer that it will create directories if they don't exist. Functions with names like `check_x` often act only as validation and tend not to manipulate things, so this is a slightly confusing break from (admittedly not super well-established) convention. I'd suggest something like `ensure_dir_path_exists` instead.

elichad commented on muspinsim/__main__.py:
```diff
@@ -24,29 +24,73 @@
+        )
         parser.add_argument(
-            "input_file",
+            "-o",
```
Another common convention is to have optional arguments in alphabetical order (aside from `-h` which is sometimes kept at the top). Makes it easier if you're looking for something specific in the help section.

elichad commented:
This could also do with some tests – maybe more integration-type tests where you model a user putting in certain command-line arguments, and then make sure outputs are written to the correct locations etc.

(sorry for the delayed review, btw)

anish-mudaraddi commented on test/test_args.py:
```diff
@@ -0,0 +1,165 @@
+                    cmd_args,
+                )
+
+                # self.assertTrue(os.path.exists("{0}/new_test.log".format(out_tmp_dir)))
```
There seems to be a bug where the log file doesn't get created when this test is run.

elichad replied:
I see commented lines like this in every test - do they all have this problem?

elichad commented on test/test_args.py:
```diff
@@ -0,0 +1,165 @@
+            with tempfile.TemporaryDirectory() as out_tmp_dir:
+                cmd_args = [
+                    "-o",
+                    "{0}".format(out_tmp_dir),
+                    "-l",
+                    "{0}/new_test.log".format(out_tmp_dir),
```
This is fine as-is, but are you familiar with [f-strings](https://realpython.com/python-f-strings/)? It's a matter of personal preference but I like them for readability. Suggestion is an example of what they'd look like in this case, you don't have to commit it.
```suggestion
                    f"{out_tmp_dir}", 
                    "-l",
                    f"{out_tmp_dir}/new_test.log",
```

elichad commented on test/test_args.py:
```diff
@@ -0,0 +1,165 @@
+    main()
+
+
+class TestCommandLineArgs(unittest.TestCase):
```
Good work adding these tests :)

anish-mudaraddi commented on muspinsim/__main__.py:
```diff
@@ -77,19 +77,32 @@ def main(use_mpi=False):
+            # in case it was just a filename given
+            # - default directory is input dir
+            except NotADirectoryError:
+                log_path = ensure_dir_path_exists(
+                    os.path.dirname("{0}/{1}".format(inp_dir, args.log_path))
+                )
```
yep that's a mistake is should just be
`log_path = os.path.join(inp_dir, args.log_path)`

anish-mudaraddi commented on muspinsim/__main__.py:
```diff
@@ -77,19 +77,32 @@ def main(use_mpi=False):
-                check_dir_path(os.path.dirname(args.log_path)),
-                os.path.basename(args.log_path),
+                ensure_dir_path_exists(os.path.dirname(log_path)),
+                os.path.basename(log_path),
```
yea that makes sense - we can just change this to
`os.path.join(os.path.dirname(log_path), os.path.basename(log_path))`

anish-mudaraddi commented on muspinsim/__main__.py:
```diff
@@ -134,12 +146,22 @@ def main(use_mpi=False):
+                    rep_fname = args.fit_report_path
+                    rep_dname = inp_dir
+
+                # default to creating it with outputs
```
i just removed this comment

anish-mudaraddi commented:
I fixed the issue with log files not being created

previously when running:
```
logging.basicConfig()
```

 if it found any existing log handlers, it would return without doing anything
 
 I added the `force` flag to ignore log handlers that `unittest` adds.

patrick-austin commented on muspinsim/__main__.py:
```diff
@@ -72,26 +127,46 @@ def main(use_mpi=False):
+            rep_dname = inp_dir
+            rep_fname = "{0}_fit_report.txt".format(inp_file_name)
+            if args.fit_report_path:
+                try:
+                    ensure_dir_path_exists(os.path.dirname(args.fit_report_path))
+                    rep_fname = os.path.basename(args.fit_report_path)
+                    rep_dname = os.path.dirname(args.fit_report_path)
+
+                # in case it was just a filename given
+                # - default directory is input dir
+                except NotADirectoryError:
+                    rep_fname = args.fit_report_path
+                    rep_dname = inp_dir
```
```suggestion
                try:            
                    # check if directory exists, if not create it
                    ensure_dir_path_exists(os.path.dirname(args.fit_report_path))
                    rep_fname = os.path.basename(args.fit_report_path)
                    rep_dname = os.path.dirname(args.fit_report_path)
                except NotADirectoryError:
                    # if fit_report_path was just a filename use inp_dir as a default
                    rep_fname = args.fit_report_path
                    rep_dname = inp_dir
```
If above change to comments accepted, make this consistent.

patrick-austin commented on muspinsim/__main__.py:
```diff
@@ -24,29 +24,73 @@
+            "input_file",
+            type=ap.FileType("r"),
+            default=None,
+            help="""filepath to muspinsim specially formatted text
+            file specifying simulation input parameters.""",
```
I find the suggested change easier to read/understand.

patrick-austin reviewed (commented):
Only have minor comments on readability of code (comments). @joelvdavies I'd value a second opinion on my suggested changes - if you think it's better as it currently is I'm happy to approve this now, or if you think something else would be more readable we can go with that. It's my first time looking at mu spin sim code so don't want my unfamiliarity with the codebase to result in pointless nitpicking.
